File: src/ProgenyImputedFamilyRef.cpp
#include <algorithm>
#include <cassert>
#include "../include/ProgenyImputedFamilyRef.h"
#include "../include/VCFFamily.h"
#include "../include/VCFProgenyImputed.h"
#include "../include/KnownFamily.h"
#include "../include/OptionSmall.h"
#include "../include/common.h"
#include "../include/RefCommon.h"

using namespace std;


//////////////////// ProgenyImputedFamilyRef ////////////////////

VCFGeno *ProgenyImputedFamilyRef::impute(const VCFSmall *orig_vcf,
								const VCFGenoBase *imputed_vcf,
								const vector<const KnownFamily *>& families,
								const vector<vector<string>>& imputed_progenies,
								const vector<vector<int>>& ref_haps,
								const OptionSmall& op) {
	
	vector<VCFImputable *>	vcfs;
	vector<VCFGeno *>	vcf_garbage;
	const size_t	L = families.size();
	for(size_t i = 0; i < L; ++i) {
		const KnownFamily	*family = families[i];
		const string&	progeny = imputed_progenies[i][0];
		STRVEC	samples = { family->get_mat(), family->get_pat(), progeny };
		auto	*vcf = VCFGeno::extract_samples(samples, orig_vcf);
		const auto	records = RefCommon::merge_family_records(imputed_vcf,
																vcf, samples);
		auto	*vcf1 = new VCFProgenyImputed(samples, records,
											  ref_haps, family->is_mat_known(),
											  op.map, 0.01, orig_vcf);
		vcfs.push_back(vcf1);
		vcf_garbage.push_back(vcf);
	}
	
	if(vcfs.empty()) {
		Common::delete_all(vcf_garbage);
		return NULL;
	}
	
	// Small VCFs are heavy to process, so it will be parallelized.
	VCFImputable::impute_VCFs(vcfs, op.num_threads);
	cout << vcfs.size()
			<< " families whose progeny is imputed have been imputed." << endl;
	
	auto	*new_vcf = VCFImputable::join(vcfs, orig_vcf->get_samples());
	Common::delete_all(vcf_garbage);
	Common::delete_all(vcfs);
	return new_vcf;
}

File: python/ProgenyImputedFamilyRef.py
from __future__ import annotations

# coding: utf-8
# ProgenyImputedFamilyRef.py
# 片親がknownで後代が一つ以上imputeされている家系を補完する

from typing import Optional

from VCF import VCFSmall
from VCFGeno import VCFGenoBase, VCFGeno
from VCFProgenyImputed import VCFProgenyImputed
import ProgenyImputedFamily
import RefCommon
from KnownFamily import *
from OptionSmall import OptionSmall


#################### ProgenyImputedFamilyRef ####################

def impute(orig_vcf: VCFSmall, phased_vcf: VCFGenoBase,
			families: list[KnownFamily], imputed_progenies: list[list[str]],
			ref_haps: list[list[int]], op: OptionSmall) -> Optional[VCFGeno]:
	vcfs: list[VCFGenoBase] = []
	for i in range(len(families)):
		family = families[i]
		progeny = imputed_progenies[i][0]
		samples = [family.mat, family.pat, progeny]
		vcf = VCFGeno.extract_samples(samples, orig_vcf);
		records = RefCommon.merge_family_records(phased_vcf, vcf, samples)
		vcf1 = VCFProgenyImputed(samples, records, ref_haps,
									family.mat_known, op.map, orig_vcf)
		vcf1.impute()
		parent = family.mat if family.mat_known else family.pat
		vcf_parent = VCFGenoBase.extract_by_samples(vcf1, [parent])
		vcfs.append(vcf_parent)
	
	if not vcfs:
		return None
	
	new_vcf = VCFGeno.join(vcfs, orig_vcf.samples)
	
	print("%d self families have been imputed." % len(families))
	return new_vcf

__all__ = ['impute']
